Crypto::Bcrypt: correct implementation

ysbaddaden opened the pull request:
Correct implementation of the Bcrypt password cipher.

Fixes:
- generates valid hashes that can be verified in other implementations and can verify hashes generated in other implementations;
- Blowfish now uses 32bits integers instead of 64bits (~800 times faster) but is still 40% slower than pure C implementations (as used by Ruby's bcrypt gem for instance);

Breaking Changes:
- Crypto::Bcrypt class now merely generates hashes;
- Introduces Ruby-like Crypto::Bcrypt::Password class for a simpler interface:

```crystal
password = Crypto::Bcrypt::Password.create("secret", cost: 12)
password == "secret" # => true
```

Notes:
- I didn't verify that the implementation is safe regarding the 2011 bug (I'm missing some test vectors for that).
- An enhancement could be to automatically select a time-based default cost, instead of a fixed one.

ysbaddaden commented:
Crystal -> Ruby:
```
>> BCrypt::Password.new("$2a$04$MECyW0PgKUWuXBCyL0SzXe1JtuGz3aMREhwZRNvrKRYcJzV1kGlLds")              
BCrypt::Errors::InvalidHash: invalid hash
```

Ruby -> Crystal:
```
Crypto::Bcrypt.verify("secret", "$2a$12$DXhAe3XzXdjMSLD2XbRjQeYoWYXrjFdeRC5xFr17c6kXeMzrSsEXm")
Invalid byte sequence in UTF-8 string (InvalidByteSequenceError)
*Exception@Exception#initialize<InvalidByteSequenceError, String>:Array(String) +31 [0]
*InvalidByteSequenceError#initialize<InvalidByteSequenceError, String>:Array(String) +6 [0]
*InvalidByteSequenceError::new<String>:InvalidByteSequenceError +89 [0]
*InvalidByteSequenceError::new:InvalidByteSequenceError +21 [0]
*CharReader#invalid_byte_sequence<CharReader>:NoReturn +10 [0]
*CharReader#decode_current_char<CharReader>:Char +157 [0]
*String#[]<String, Int32>:Char +731 [0]
*Crypto::Blowfish#next_word<Crypto::Blowfish, String, Pointer(Int32)>:Int32 +135 [0]
*Crypto::Blowfish#salted_expand_key<Crypto::Blowfish, String, String>:Int32 +709 [0]
*Crypto::Bcrypt::setup<String, String, String>:Crypto::Blowfish +64 [0]
*Crypto::Bcrypt::bcrypt<String, String, String>:String +60 [0]
*Crypto::Bcrypt::verify<String, String>:Bool +258 [0]
```

jhass commented:
Yeah, it should decode to a slice, not to a string.

technorama commented:
Test vectors are available in #1494.  Most are commented out due to failures in the current implementation of Bcrypt.  

Is there a reason we don't just import bcrypt.c?  Making a custom cryptography library is often error prone.

asterite commented:
Does this need a rebase now that Base64 is fixed?

jhass commented:
It uses a separate (probably broken) Base64 implementation.

ysbaddaden commented:
I eventually fixed the Bcrypt implementation.

jhass commented:
Awesome! Could you pull in the vectors from #1494 ?

ysbaddaden commented:
Already did.

ysbaddaden commented on src/crypto/bcrypt.cr:
```diff
@@ -1,153 +1,119 @@
 
-  MIN_COST = 4
-  MAX_COST = 63
   DEFAULT_COST = 10
```
A default cost of 10 may be a very bad default now: it takes 100ms (60ms in C) to hash a password on my laptop. Maybe a cost of 12 (400ms, but 240ms in C) would be a better default.

jhass replied:
I'd trust your judgement there.

jhass commented:
Ah, I guess I didn't recognize them in their expanded form :P

ysbaddaden commented on spec/std/crypto/bcrypt_spec.cr:
```diff
@@ -1,50 +1,38 @@
+    { 6,  "~!@#$%^&*()      ~!@#$%^&*()PNBFRD", "fPIsBO8qRqkjj273rfaOI.", "HtSV9jLDpTbZn782DC6/t7qT67P6FfO" },
+    { 8,  "~!@#$%^&*()      ~!@#$%^&*()PNBFRD", "Eq2r4G/76Wv39MzSX262hu", "zPz612MZiYHVUJe/OcOql2jo4.9UxTW" },
+    { 10, "~!@#$%^&*()      ~!@#$%^&*()PNBFRD", "LgfYWkbzEvQ4JakH7rOvHe", "0y8pHKF9OaFgwUZ2q7W2FFZmZzJYlfS" },
+  ]
```
Actually I miss these vectors:

```crystal
# latin-1 POUND SIGN
{ 5, "\xa3", "CCCCCCCCCCCCCCCCCCCCC.", "BvtRGGx3p8o0C5C36uS442Qqnrwofrq" },

# utf-8 POUND SIGN
{ 5, "\xc2\xa3", "CCCCCCCCCCCCCCCCCCCCC.", "CAzSxlf0FLW7g1A5q7W/ZCj1xsN6A.e" }

# add 8-bit unicode test as well; to verify PY3 encodes it as UTF-8.
{ 5, "\u00A3", "CCCCCCCCCCCCCCCCCCCCC.", "CAzSxlf0FLW7g1A5q7W/ZCj1xsN6A.e" },
```

But only the last one is valid in Crystal (`\u00A3`) since `\x00` isn't expanded as a hexadecimal char.

jhass replied:
Ugly construct (`Array` is missing `to_slice`), but `String.new(Slice.new(UInt8[0xa3].to_unsafe, 1))` if you need them as a String,  though it would probably good to accept the Slice and maybe even Array too.

ysbaddaden commented:
I upgraded the default cost to 11, which we should upgrade to 12 once the implementation is on par with pure C implementations. I also fixed some validation tests and added a test for the sign expansion (high 8bit) bug.

Looks good enough for me.

jhass commented:
Awesome work <3

ysbaddaden commented:
I have no idea how to speed things up, thought. Maybe using Slices(Int32) instead of Array(Int32) for the P and S arrays in Crypto::Blowfish would help, but I'm not sure how to initialize them (maybe when we dup S and P).
